mt.c

```diff
@@ -612,16 +611,26 @@ void execsh(void) {
       die("who are you?\n");
   }
 
-  if ((sh = getenv("SHELL")) == NULL)
-    sh = (pw->pw_shell[0]) ? pw->pw_shell : shell;
+  // Find user's preferred shell. This will become $SHELL.
+  char *sh;
+  if (shell != NULL) {
+    sh = shell;
+  } else if ((sh = getenv("SHELL")) != NULL) {
+  } else if (pw->pw_shell[0]) {
+    sh = pw->pw_shell;
+  } else {
+    sh = "/bin/sh";
+  }
 
-  if (opt_cmd)
+  // The program to execute is $SHELL unless explicitly specified.
```

Comment:
I would say `The program to execute is the shell unless ...` because above you handled finding a shell that is broader than `$SHELL`.